File: ts_sdk/src/claude_hook_events/event-types.ts
/**
 * Hook event types and grouping constants for Claude Code hooks.
 */

/**
 * Types of hook events from Claude Code.
 * Based on Claude Code hooks documentation:
 * https://code.claude.com/docs/en/hooks
 */
export enum HookEventType {
  // Session lifecycle events
  SESSION_START = 'SessionStart',
  SESSION_END = 'SessionEnd',

  // User interaction events
  USER_PROMPT_SUBMIT = 'UserPromptSubmit',
  NOTIFICATION = 'Notification',

  // Tool events
  PRE_TOOL_USE = 'PreToolUse',
  POST_TOOL_USE = 'PostToolUse',
  POST_TOOL_USE_FAILURE = 'PostToolUseFailure',
  PERMISSION_REQUEST = 'PermissionRequest',

  // Agent stop/start events
  STOP = 'Stop',
  STOP_FAILURE = 'StopFailure',
  SUBAGENT_START = 'SubagentStart',
  SUBAGENT_STOP = 'SubagentStop',

  // Agent teams events
  TEAMMATE_IDLE = 'TeammateIdle',
  TASK_CREATED = 'TaskCreated',
  TASK_COMPLETED = 'TaskCompleted',

  // Configuration events
  CONFIG_CHANGE = 'ConfigChange',
  INSTRUCTIONS_LOADED = 'InstructionsLoaded',

  // Worktree events
  WORKTREE_CREATE = 'WorktreeCreate',
  WORKTREE_REMOVE = 'WorktreeRemove',

  // Compaction events
  PRE_COMPACT = 'PreCompact',
  POST_COMPACT = 'PostCompact',

  // MCP elicitation events
  ELICITATION = 'Elicitation',
  ELICITATION_RESULT = 'ElicitationResult',

  // File system events
  CWD_CHANGED = 'CwdChanged',
  FILE_CHANGED = 'FileChanged',
}

/**
 * Hook events that don't use matchers (always fire on every occurrence)
 */
export const HOOK_EVENTS_NO_MATCHER = [
  HookEventType.USER_PROMPT_SUBMIT,
  HookEventType.STOP,
  HookEventType.TEAMMATE_IDLE,
  HookEventType.TASK_COMPLETED,
  HookEventType.WORKTREE_CREATE,
  HookEventType.WORKTREE_REMOVE,
  HookEventType.CWD_CHANGED,
  HookEventType.FILE_CHANGED,
] as const;

/**
 * Hook events that use matchers
 */
export const HOOK_EVENTS_WITH_MATCHER = [
  HookEventType.PRE_TOOL_USE,
  HookEventType.POST_TOOL_USE,
  HookEventType.POST_TOOL_USE_FAILURE,
  HookEventType.PERMISSION_REQUEST,
  HookEventType.SESSION_START,
  HookEventType.SESSION_END,
  HookEventType.NOTIFICATION,
  HookEventType.SUBAGENT_START,
  HookEventType.SUBAGENT_STOP,
  HookEventType.STOP_FAILURE,
  HookEventType.PRE_COMPACT,
  HookEventType.POST_COMPACT,
  HookEventType.CONFIG_CHANGE,
  HookEventType.INSTRUCTIONS_LOADED,
  HookEventType.ELICITATION,
  HookEventType.ELICITATION_RESULT,
] as const;

/**
 * All available hook events
 */
export const ALL_HOOK_EVENTS = [...HOOK_EVENTS_NO_MATCHER, ...HOOK_EVENTS_WITH_MATCHER] as const;

/**
 * Default hooks to listen to — excludes worktree create event (which would replace the default behavior)
 */
export const DEFAULT_LISTENED_HOOKS = ALL_HOOK_EVENTS.filter(
  (e) => e !== HookEventType.WORKTREE_CREATE
);

File: flow_sdk/claude_hook_events/event_types.py
"""Hook event type enum for Claude Code hooks."""

from __future__ import annotations

from flow_sdk._compat import StrEnum


class HookEventType(StrEnum):
    """Available Claude Code hook event types."""

    # User interaction hooks
    USER_PROMPT_SUBMIT = "UserPromptSubmit"
    """Triggered when user submits a prompt."""

    # Tool lifecycle hooks
    PRE_TOOL_USE = "PreToolUse"
    """Triggered before a tool is executed. Can block or modify input."""

    POST_TOOL_USE = "PostToolUse"
    """Triggered after a tool completes. Can run cleanup/validation."""

    POST_TOOL_USE_FAILURE = "PostToolUseFailure"
    """Triggered after a tool fails."""

    # Session hooks
    STOP = "Stop"
    """Triggered when Claude finishes responding."""

    STOP_FAILURE = "StopFailure"
    """Triggered when a turn ends due to an API error."""

    NOTIFICATION = "Notification"
    """Triggered when Claude needs user input."""

    SESSION_START = "SessionStart"
    """Triggered when session starts or resumes."""

    SESSION_END = "SessionEnd"
    """Triggered when session ends."""

    INSTRUCTIONS_LOADED = "InstructionsLoaded"
    """Triggered when CLAUDE.md or .claude/rules/*.md files are loaded."""

    # Context hooks
    PRE_COMPACT = "PreCompact"
    """Triggered before context compaction."""

    POST_COMPACT = "PostCompact"
    """Triggered after context compaction completes."""

    # Subagent hooks
    SUBAGENT_START = "SubagentStart"
    """Triggered when a subagent starts."""

    SUBAGENT_STOP = "SubagentStop"
    """Triggered when a subagent completes its task."""

    # Permission hooks
    PERMISSION_REQUEST = "PermissionRequest"
    """Triggered when a tool requests permission."""

    # Agent teams events
    TEAMMATE_IDLE = "TeammateIdle"
    """Triggered when a teammate is idle."""

    TASK_COMPLETED = "TaskCompleted"
    """Triggered when a task is completed."""

    # Configuration events
    CONFIG_CHANGE = "ConfigChange"
    """Triggered when configuration changes."""

    # Worktree events
    WORKTREE_CREATE = "WorktreeCreate"
    """Triggered when a worktree is created."""

    WORKTREE_REMOVE = "WorktreeRemove"
    """Triggered when a worktree is removed."""

    # MCP elicitation events
    ELICITATION = "Elicitation"
    """Triggered when an MCP server requests user input during a tool call."""

    ELICITATION_RESULT = "ElicitationResult"
    """Triggered after a user responds to an MCP elicitation."""

    # File system events
    CWD_CHANGED = "CwdChanged"
    """Triggered when the current working directory changes."""

    FILE_CHANGED = "FileChanged"
    """Triggered when a file is created, modified, or deleted."""
